semantic-ast/src/Marshal/JSON.hs

```diff
@@ -0,0 +1,116 @@
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Marshal.JSON
+( MarshalJSON(..)
+) where
+
+import Data.Aeson as Aeson
+import Data.List.NonEmpty (NonEmpty)
+import GHC.Generics
+import Data.Text (Text)
+import qualified Data.Text as Text
+-- import Marshal.Examples
+
+-- Test datatype that will go away: this is just to get us started!
+data Bar a = Bar
+  { ann :: a
+  , foo :: Text
+  } deriving (Eq, Show, Generic1)
+  -- Aeson requires that the datatypes derive Generic
+  -- we want Generic1 because we can represent types of kind * -> *
+
+-- Thinking about the shape of Bar:
+-- it is a Datatype, meaning it will require M1 D
+-- it has a data constructor, so it will require C1
+-- it is a product type, so it will require :*:
+-- it has record selector fields so it will require S1
+-- is has a parameter so it will require Par1
+-- it has a constant, foo :: Text, so it will require K1 R
+
+-- Serialize unmarshaled ASTs into JSON representation by auto-deriving Aeson instances generically
+-- Given some shape t with polymorphic type a such that a has a ToJSON instance, we can produce a JSON value
+-- in the default definition, we add further constraints: 
+-- - Generic1 t - the type constructor will be a sum, product, leaf etc. and will require a Generic1 instance
+-- - GMarshalJSON (Rep1 t) -- the type constructor should be a Rep1, which is an associated type (ie., an open type family affiliated with a class, and thus it's like a function for types, in this case the Haskell language datatypes used in the AST)
+class MarshalJSON t where
+  marshal :: (ToJSON a) => t a -> Value
+  default marshal :: ( Generic1 t, GMarshalJSON (Rep1 t), ToJSON a) => t a -> Value
+  marshal = gmarshal . from1
```

Comment:
Marshaling, in this case, means that given some shape `t` with polymorphic type `a` such that `a` has a `ToJSON` instance, we can produce a JSON `Value`.  

In the default definition, we add further constraints: 
- `Generic1 t` - the type constructor will be a sum, product, leaf, etc. and will require a `Generic1` instance
- `GMarshalJSON (Rep1 t)` - the type constructor should be a `Rep1`, which is an associated type (ie., an open type family affiliated with a class, and thus it's like a function for types (in this case the Haskell language datatypes used in the AST).

Reply:
The latter constraint more specifically means that the generic representation of t—`Rep1 t`—must have a `GMarshalJSON` instance.